### PR TITLE
feat(flopy plugins)

### PR DESCRIPTION
**Pull request by spaulins-usgs**
FloPy plugins are FloPy-integrated components that modify MODFLOW-6 functionality using the modflowapi library. The FloPy plugins architecture streamlines the process of modifying MODFLOW-6 functionality by:

* Providing template generators to help rapidly set up your plugin’s interface
* Automatically loading, saving, and running FloPy plugins as part of a MODFLOW-6 simulation
* Providing an interface that allows multiple FloPy plugins to run together
* Providing support for running FloPy plugins installed as other Python packages

### CODE REVIEW

File: pyproject.toml
Comment (mwtoews) [inline pyproject.toml:78]: is it possible to not pin these to git develop versions? These pip dependencies are tricky to use, for instance, someone may have already done `pip install -e .` for (e.g.) a local develop clone of xmipy, but this would get uninstalled/replaced with a non-editable copy of the same repo. Ideally it should be pinned to a minimum version released on PyPI, which would normally leave a local develop clone alone if it is up-to-date.

Comment (spaulins-usgs) [reply pyproject.toml:78]: @mwtoews, this feature is dependent on versions of xmipy and modflowapi that have not been released yet.  Once those versions have been released these dependencies can be updated as you suggested.

Comment (spaulins-usgs) [reply pyproject.toml]: @w-bonelli, good idea.  Changed the PR so that FloPy now lets modflowapi fulfill the xmipy dependency.

Comment (mwtoews) [reply pyproject.toml]: A better workaround is to unpin the explicit development version here and do a separate `pip install modflowapi @ ...` command in the CI workflow before installing `flopy[api]`. In related developments, [xmipy v1.3.0](https://github.com/Deltares/xmipy/releases/tag/v1.3.0) was released yesterday.

Comment (spaulins-usgs) [reply pyproject.toml]: @mwtoews, I apologize in advance since I am still learning about the .toml file and these CI workflows.  I do not understand why the develop version of modflowapi should be removed from pyproject.toml.  Shouldn't the pyproject.toml file reflect what needs to be installed for FloPy to function properly?

Note that the next modflowapi release will likely happen soon, maybe even before this FloPy plugins PR gets approved.  When it does I can remove all modflowapi develop version dependencies from pyproject.toml and all the workflows, which will solve the problem.  This is just a short-term solution so that I can get this feature into review without having to wait for the next release of modflowapi.

Comment (mwtoews) [reply pyproject.toml]: The reason why the develop version of modflowapi should be unpinned from pyproject.toml is that it will overwrite any other local version of modflowapi, which I find disruptive (since I probably already have a develop version installed). It is overwritten since the develop version has no version number to compare. My suggestion is analogous to xmipy to first [install xmipy/modflowapi in the CI script](https://github.com/MODFLOW-USGS/modflowapi/blob/a8e241df1def5899ccbf22368eddc76da0d7a60c/.github/workflows/ci.yml#L124-L126), e.g. before `pip install .[api]`. The unpinned version should look like [this pyproject.toml](https://github.com/MODFLOW-USGS/modflowapi/blob/a8e241df1def5899ccbf22368eddc76da0d7a60c/pyproject.toml#L37) with a helpful comment. A runtime check is a bit more difficult for me to suggest, as I don't know what features are being used in the develop version, but it might be not necessary if a release is immanent. Further clarification is always welcome!

File: examples/Notebooks/dev/flopy3_dev_mf6_packaging_distributing_flopy_plugins.ipynb
Comment (langevin-usgs) [inline examples/Notebooks/dev/flopy3_dev_mf6_packaging_distributing_flopy_plugins.ipynb:10]:
```suggestion
    "FloPy plugins can be installed as python packages.  FloPy will automatically detect plugins in other python packages by using python package metadata.\n",
```

I don't think we want to hyphenate plugin.

File: examples/Notebooks/dev/flopy3_dev_mf6_packaging_distributing_flopy_plugins.ipynb
Comment (langevin-usgs) [inline examples/Notebooks/dev/flopy3_dev_mf6_packaging_distributing_flopy_plugins.ipynb:14]: This is confusing.  Is this extension or abbreviation?  Not sure what to do about this.  I see you've adopted extension fairly widely.  And maybe this is okay, but it might be good to say the extension is an abbreviation for the plugin that users will see.  It is similar to RIV for River or SFR for Stream Flow Routing in MODFLOW terminology, for example.

File: examples/Notebooks/dev/flopy3_dev_mf6_packaging_distributing_flopy_plugins.ipynb
Comment (langevin-usgs) [inline examples/Notebooks/dev/flopy3_dev_mf6_packaging_distributing_flopy_plugins.ipynb:1]: How does someone remove a flopy plugin?  Is this mentioned anywhere?

File: examples/Tutorials/modflow6plugin/tutorial01_mf6_plugin.py
Comment (langevin-usgs) [inline examples/Tutorials/modflow6plugin/tutorial01_mf6_plugin.py:1]: Similar comments in these .py files about MODFLOW-6 hyphenation, use of BMI terminology.  Do we need to rethink use of "dll" as it is OS specific?  

What does rvp stand for?  River Plugin?  If so should you say rvp plugin?  I know these are little details, but I think they matter.

File: examples/Tutorials/modflow6plugin/tutorial02_mf6_plugin.py
Comment (langevin-usgs) [inline examples/Tutorials/modflow6plugin/tutorial02_mf6_plugin.py:53]: Here you use abbreviation and name, which is different from earlier descriptions.

File: examples/Tutorials/modflow6plugin/tutorial02_mf6_plugin.py
Comment (langevin-usgs) [inline examples/Tutorials/modflow6plugin/tutorial02_mf6_plugin.py:135]: I think we need something like the following.  Not sure where it should go.

MODFLOW 6 has a special type of stress package, called the Application Programming Interface (API) Package.  This package does nothing by itself, but it does provide empty arrays that can be filled from another program that is accessing it through the MODFLOW API.  In the context described here, the plugin transfers stress package information, calculated with Python, into the MODFLOW API Package while MODFLOW is running.  MODFLOW then proceeds with this information and includes it in solution of the matrix equations for a model.

File: flopy/mf6/modflow/mfsimulation.py
Comment (langevin-usgs) [inline flopy/mf6/modflow/mfsimulation.py:1]: It seems like there is a lot of new flopy plugin code that was added here.  Would it be possible to move this into a plugin folder somewhere as most of it is not needed for the dominant regular flopy use case.

File: flopy/mf6/utils/flopy_plugins/plugin_interface.py
Comment (langevin-usgs) [inline flopy/mf6/utils/flopy_plugins/plugin_interface.py:174]: Maybe there is good reason for this, but this does not seem like a good place for replicated low-level GWF NPF calculation routines.

**Review by langevin-usgs — CHANGES_REQUESTED** [review]
Hi @spaulins-usgs, lots of really great stuff here.  This is some tremendous new functionality.  I tried to go through it all and make some comments.  I added quite a few, but hopefully they will be helpful.  Happy to discuss further.

File: .docs/code.rst
Comment (spaulins-usgs) [reply .docs/code.rst:203]: This section is referring to the plugin_interface.py and plugin_template.py files, which are for the plugin developer, not the user.  However, I agree that there should be a section here for the plugin user.  I therefore added a section describing how to use FloPy plugins and reference the documentation of the “package interface” files of the RVC and RVP plugins (the two plugins included with FloPy) to help the user get started.

The instructions are up to date with the recent changes I made.

File: .docs/code.rst
Comment (spaulins-usgs) [reply .docs/code.rst]: Changed

File: examples/Notebooks/dev/flopy3_dev_mf6_BuildStressPluginAPIPkgSupport.ipynb
Comment (spaulins-usgs) [reply examples/Notebooks/dev/flopy3_dev_mf6_BuildStressPluginAPIPkgSupport.ipynb]: Changed

Comment (spaulins-usgs) [reply examples/Notebooks/dev/flopy3_dev_mf6_BuildStressPluginAPIPkgSupport.ipynb]: That is the file name not the folder name.  For clarity I reworded the sentence describing this.

File: examples/Notebooks/dev/flopy3_dev_mf6_BuildStressPluginAPIPkgSupport.ipynb
Comment (spaulins-usgs) [reply examples/Notebooks/dev/flopy3_dev_mf6_BuildStressPluginAPIPkgSupport.ipynb:1]: Hyphens removed

Changed all references to API, including in the code itself (renaming some code interfaces).

Changed npc to mfp  (My Flopy Plugin) and wpf to wpm (Well Package Modifier)

File: examples/Notebooks/dev/flopy3_dev_mf6_packaging_distributing_flopy_plugins.ipynb
Comment (spaulins-usgs) [reply examples/Notebooks/dev/flopy3_dev_mf6_packaging_distributing_flopy_plugins.ipynb]: Fixed this everywhere.

File: examples/Notebooks/dev/flopy3_dev_mf6_packaging_distributing_flopy_plugins.ipynb
Comment (spaulins-usgs) [reply examples/Notebooks/dev/flopy3_dev_mf6_packaging_distributing_flopy_plugins.ipynb:1]: Added a sentence explaining that to uninstall the FloPy plugin all you have to do is uninstall the python package with the plugin in it.

File: examples/Tutorials/modflow6plugin/tutorial01_mf6_plugin.py
Comment (spaulins-usgs) [reply examples/Tutorials/modflow6plugin/tutorial01_mf6_plugin.py:1]: I changed “dll” to the more OS-neutral “lib” or “library” everywhere.

Added that rvp stands for “RiVer Plugin”

File: examples/Tutorials/modflow6plugin/tutorial02_mf6_plugin.py
Comment (spaulins-usgs) [reply examples/Tutorials/modflow6plugin/tutorial02_mf6_plugin.py:53]: I change “extension” to “abbreviation” everyone for consistency.

File: flopy/mf6/utils/flopy_plugins/plugin_interface.py
Comment (spaulins-usgs) [reply flopy/mf6/utils/flopy_plugins/plugin_interface.py]: Removed, as commented above.

File: .docs/tutorials.rst
Comment (jlarsen-usgs) [inline .docs/tutorials.rst:54]: It'd be nice to have a tutorial on how to override functionality in an existing package. The current set of tutorials shows how to use an existing plugin and how to develop one in the API package. Modifying existing package behavior is an important topic.

**Review by jlarsen-usgs — CHANGES_REQUESTED** [review]
Not really convinced that the name os this should FloPy Plugins. This is not general use plugin code that allows any user to create python packages that modifies the behavior of FloPy. Instead this is plugin code that allows users to develop python packages and api callback containers for modflow 6 that then can be used as plugins with flopy.

For example, with this implementation I couldn't develop specialized plugin code that could handle reading modflow-owhm or modify pygsflow to be an optional plugin/extension for flopy. Some general plugin hooks to make things like this possible would be really beneficial. I could imagine in the future we could then have something like flopy[core] (which is the core gwf and gwt functionality), and as modflow 6 grows additional supported components (e.g. support for the surface water model being developed) could be installed as plugins to core functionality. This is similar to how "ray" packages their different functionalities. 

The actual plugin code is really buried in flopy.mf6.utils.plugins, is there a better location for this and if we develop generalized plugin hooks too, we should think about where it should live?

I've left some comments where I think the code could be simplified or flopy is duplicating work that is already being done by modflowapi.

Overall, the Modflow 6 plugin system you've developed seems really useful and I think it'll be a nice addition to flopy and modflowapi! Let me know if you have any questions about my review comments.

**Comment by langevin-usgs** [comment]
@jlarsen-usgs, these are great comments.  Lots to consider.  Makes me wonder if flopy is the right place to put these "plugins".  They might be more aptly named plugins if they were included in modflowapi instead.  That way the "plugin" term in flopy could be leveraged for the uses cases that you mention.

File: pyproject.toml
Comment (spaulins-usgs) [reply pyproject.toml]: Now that modflowapi version 0.1.0 has been released I have removed the develop version pip dependencies from flopy.  This should resolve the issue.

File: flopy/mf6/utils/flopy_plugins/plugin_interface.py
Comment (jdhughes-usgs) [inline flopy/mf6/utils/flopy_plugins/plugin_interface.py:109]: Looks like this returns a single value. If true couldn't this function be `value_from_cellid()`.

`grid_array` can just be `array` or `arr`

Why would it ever be another `modelgrid.grid_type`? If you want to provide protection it should issue an error if it another `grid_type`.

File: flopy/mf6/utils/flopy_plugins/plugin_runner.py
Comment (jdhughes-usgs) [inline flopy/mf6/utils/flopy_plugins/plugin_runner.py:36]: `get_api_library`

File: flopy/mf6/modflow/mfsimulation.py
Comment (jdhughes-usgs) [inline flopy/mf6/modflow/mfsimulation.py:1617]: I think we need to think about if we want flopy to open a separate file for writing information. Users can always specify `print_input` and `print_budget` to have information written to the mf6 lst file.

**Review by jdhughes-usgs — COMMENTED** [review]
There is alot here to digest. I am still not convinced this belongs in flopy. It belongs somewhere but maybe not in flopy.

I like Josh's idea that a flopy plugin would extend flopy functionality. I could see a flopy plugin providing input and output support for MODFLOW-OWHM or MODFLOW-USG. That way it would be up to the person developing the plugin to support it and flopy could support what we want it to. We could think about a version of flopy (flopy4) that only supports mf6 and base functionality (modelgrid, geoproocessing, plotting, etc) and move old MODFLOW, MT3D, SEAWAT, etc functionality out as a plugin (this is an idea Chris brought up).

What you are calling a "flopy plugin" is providing a pre-canned way to use the API that seems like flopy. Josh's additions to the modflowapi package already helps alot in allowing users to interact with the API in a flopy way. So maybe "flopy plugins" should just be additional helper functionality in the modflowapi package.  

I think more thought needs to be given to naming. Function names and variables should provide users with an unambiguous idea of what is going on.

I don't like that flopy dfns are in the repo. It seems like an opportunity for confusion over who owns dfns.

I am not completely clear on the difference between the rvc and rvp examples. They do seem overly complicated. I think one of the cases should be using a standard river package with an aux variable that defines either the inflow or outflow conductance and the code picks the appropriate one depending on the flow direction. Maybe I missed something.

**Comment by spaulins-usgs** [comment]
I am closing this PR since Flopy Plugins (now Flopy API Plugins) is now going to be its own library.

**Pull request closed without merging**
